packages/crypto/src/crypto.rs

```diff
@@ -0,0 +1,183 @@
+use k256::{
+    ecdsa::signature::{DigestVerifier, Signature as _}, // traits
+    ecdsa::{Signature, VerifyingKey},                   // type aliases
+};
+use sha2::Digest; // trait
+
+use crate::errors::{CryptoError, CryptoResult};
+use crate::identity_digest::Identity256;
+
+pub fn secp256k1_verify(
+    message_hash: &[u8],
+    signature: &[u8],
+    public_key: &[u8],
+) -> CryptoResult<bool> {
+    // Already hashed, just build Digest container
```

Comment:
Let's check the pubkey len and prefix here

Reply:
Already added tests at the imports / native api level.

Reply:
Will see to extend these.

Reply:
The tests are good. But I want to control certain kinds of checks and maybe have specific error messages for them. What we need to check here is:

1. `message_hash.len() == 32`, otherwise `Identity256::new().chain(message_hash)` crashes.
2. `signature.len() == 64` to ensure we get a signature in fixed length format, not something else like DER
3. `public_key.len() == 65 && public_key[0] == 0x04` or `public_key.len() == 33 && public_key[0] == 0x03` or `public_key.len() == 33 && public_key[0] == 0x02` to ensure we get one of the three supported formats, not anything else.

Reply:
VM interface above guarantees (through region checks) limits are not exceeded. And, implementation below (i.e. `k256` functionality) must be robust enough, to handle shorter and / or broken data.

But, we can add those tests, as you wish. Another point to consider, though, is that `k256` error messages are deliberately opaque, to avoid data leaking attacks. See https://github.com/RustCrypto/traits/blob/master/signature/src/error.rs#L10-#L18.

Honestly, I would add as little extra code / checks as possible.

Reply:
My proposal is:

We can add a check for 1. above (to avoid crashes), and let signs and pubkeys alone.
And, I'll add some local tests here, to document this, along with some comments.

Reply:
The length limit checks pretect against resource consumtion and are always upper limits. When you pass a message_hash of 31 bytes, you crash the VM.

Also this implementation should do all necessary checks independent how it is used, such that you get the exact same behaviour for unit tests, in the VM and any future use case.

`k256` checking the inputs is good. But we only want to accept a subset of the accepted inputs of `k256`.

> k256 error messages are deliberately opaque

Right. This is good and bad at the same time. When the basic format checks from above fail, a contract developer did something wrong in the logic and should be guided with a helpful error message.

Reply:
I looked into BB'06 (https://en.wikipedia.org/wiki/Daniel_Bleichenbacher#BB'06_attack:_signature_forgery_attack_against_the_RSA_PKCS#1_signature_standard) and from the description of the attack, I don't see how this is related to leakage of failure information. But independent of BB'06, I agree in general that no verification failure information should be leaked through error massages as a safety principle.

However, I don't think the basic format checks mentioned above fall in this cathegory as they simply check that the input data is encoded correctly.

Reply:
> However, I don't think the basic format checks mentioned above fall in this category as they simply check that the input data is encoded correctly.

OK. Will add checks for signature / pubkey lengths, and pubkey prefix.